HIVE-29177: Implement default Catalog selection

zhangbutao opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://cwiki.apache.org/confluence/display/Hive/HowToContribute
  2. Ensure that you have created an issue on the Hive project JIRA: https://issues.apache.org/jira/projects/HIVE/summary
  3. Ensure you have added or run the appropriate tests for your PR: 
  4. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP]HIVE-XXXXX:  Your PR title ...'.
  5. Be sure to keep the PR description updated to reflect all changes.
  6. Please write your PR title to summarize what this PR proposes.
  7. If possible, provide a concise example to reproduce the issue for a faster review.

-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->

**TODO: Add more test and cover more api.**

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->


### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description, screenshot and/or a reproducable example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Hive versions or within the unreleased branches such as master.
If no, write 'No'.
-->

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->

deniskuzZ commented:
is the PR ready for review? 
cc @difin - once merged, we should probably update the RestCatalog client to use selected catalog configs

zhangbutao commented:
> is the PR ready for review? cc @difin - once merged, we should probably update the RestCatalog client to use selected catalog configs

This PR involves many database interfaces. Some of the interfaces I've marked with TODO tags and need a double check to avoid potential issues that could be hard to troubleshoot in future development.  But I welcome reviews of this PR during the period when I'm optimizing the code.

Additionally, this PR only addresses the interfaces for switching catalogs and databases, and does not cover interfaces related to `cat.db.table`. Therefore, even if the catalog and database are switched, the current client will still query tables from the default `hive `catalog. The code related to `cat.db.tbl` (part of the work in HIVE-29177) will also involve many table-related interfaces, which presents certain refactoring challenges.

If I have enough time in the near future, I might write down some thoughts on federated catalogs, including typical use cases and the code refactoring challenges involved. I hope this will attract more developers to join the discussion.

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/ddl/database/create/CreateDatabaseDesc.java:
```diff
@@ -45,13 +46,14 @@ public class CreateDatabaseDesc implements DDLDesc, Serializable {
+  public CreateDatabaseDesc(String catalogName, String databaseName, String comment, String locationUri, String managedLocationUri,
       boolean ifNotExists, Map<String, String> dbProperties) {
-    this(databaseName, comment, locationUri, managedLocationUri, ifNotExists, dbProperties, "NATIVE", null, null);
+    this(catalogName, databaseName, comment, locationUri, managedLocationUri, ifNotExists, dbProperties, "NATIVE", null, null);
```
for the external catalogs `dbtype` should be "REMOTE", right? It seems that we always hardcode "NATIVE" and below condition never returns true
````
 if (dbtype != null && dbtype.equalsIgnoreCase("REMOTE")) {
````

zhangbutao replied:
> should we introduce a flag like native/non-native similar to tables?

This can be done when we implement specific catalogs (JDBC catalog & Iceberg catalog, etc.).

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/ddl/table/lock/show/ShowLocksOperation.java:
```diff
@@ -160,6 +160,8 @@ private ShowLocksResponse getLocksForNewFormat(HiveLockManager lockMgr) throws H
       throw new HiveException("New lock format only supported with db lock manager.");
     }
 
+    // TODO catalog. Need to add catalog into ShowLocksRequest. But ShowLocksRequest doesn't have catalog field.
```
see https://github.com/apache/hive/pull/6088/files#r2454590955

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java:
```diff
@@ -1297,7 +1303,12 @@ public void renamePartition(Table tbl, Map<String, String> oldPartSpec, Partitio
 
-  // TODO: this whole path won't work with catalogs
+  /**
+   * When you call this method, you need to ensure that the catalog has been set in the db object.
```
why not have fallback similar to https://github.com/apache/hive/pull/6088/files#diff-9d6bc2f1a2972d844b3107f29bc1731af7adf6f4c922e08904345c5ae88cd727R1443

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java:
```diff
@@ -1914,6 +1942,16 @@ public Table apply(org.apache.hadoop.hive.metastore.api.Table table) {
     }
   }
 
+  public List<Table> getTableObjects(String catName, String dbName, String pattern, TableType tableType) throws HiveException {
```
@dengzhhu653 is it ok or we should use some request object like Database?

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/metadata/HiveUtils.java:
```diff
@@ -543,4 +546,10 @@ public static String getLowerCaseTableName(String refName) {
   }
+
+  public static String getCurrentCatalogOrDefault(Configuration conf) {
+    return SessionState.get() != null ?
```
you could try Optional.ofNullable

zhangbutao replied:
Fixed in https://github.com/apache/hive/pull/6088/commits/00a75f9fe1d13cc9e49a00a3cd2b598dadef1f0b

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/parse/BaseSemanticAnalyzer.java:
```diff
@@ -1926,6 +1960,19 @@ protected Database getDatabase(String dbName, boolean throwException) throws Sem
     return database;
   }
 
+  protected Database getDatabase(String catalogName, String dbName, boolean throwException) throws SemanticException {
```
try reuse code. you could call new method within the old one without catalogName

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/parse/repl/load/message/CreateDatabaseHandler.java:
```diff
@@ -56,22 +56,23 @@ public List<Task<?>> handle(Context context)
     Database db = metaData.getDatabase();
     String destinationDBName =
         context.dbName == null ? db.getName() : context.dbName;
+    String destinationCatalogName = db.getCatalogName(); // TODO catalog. Need to double check the catalog here. Depend on HIVE-29278
```
cc @harshal-16

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/processors/HiveCommand.java:
```diff
@@ -86,6 +86,8 @@ public static HiveCommand find(String[] command, boolean findOnlyForTesting) {
       } else if (command.length > 1 && "show".equalsIgnoreCase(command[0]) &&
           "processlist".equalsIgnoreCase(command[1])) {
         return PROCESSLIST;
+      } else if(command.length > 1 && "set".equalsIgnoreCase(command[0]) && "catalog".equalsIgnoreCase(command[1])) {
```
space

deniskuzZ commented on ql/src/test/queries/clientnegative/lockneg_try_lock_cat_db_in_use.q:
```diff
@@ -0,0 +1,9 @@
+set hive.lock.numretries=0;
+set hive.support.concurrency=true;
```
that is only needed for ACID tables when DBTxnManager is used

harshal-16 commented on ql/src/java/org/apache/hadoop/hive/ql/exec/repl/incremental/IncrementalLoadTasksBuilder.java:
```diff
@@ -264,8 +265,9 @@ private Task<?> dbUpdateReplStateTask(String dbName, String replState, Task<?> p
-    AlterDatabaseSetPropertiesDesc alterDbDesc = new AlterDatabaseSetPropertiesDesc(dbName, mapProp,
-        new ReplicationSpec(replState, replState));
+    // TODO catalog. Need to double check the actual catalog here. Depend on HIVE-29278
+    AlterDatabaseSetPropertiesDesc alterDbDesc = new AlterDatabaseSetPropertiesDesc(HiveUtils.getCurrentCatalogOrDefault(conf),
```
Oh, then I think we will need to update the incremental dump filter also to make it consistent with the load.

dengzhhu653 commented:
The biggest flaw is we cannot have two same db name/table name among catalogs, this might be a question in the future.
For cross-catalog queries e.g, db1.name1 join db1.name1 in this case, CatalogAwarenessMetastoreClient mightn't be able to figure out whether the all db1.name1 tables belong to the same catalog or not.
I don't have an elegant solution yet, maybe we can add a limit that the db is unique across the catalogs, which as we do in the Hive catalog.

zhangbutao commented on ql/src/java/org/apache/hadoop/hive/ql/ddl/database/drop/DropDatabaseAnalyzer.java:
```diff
@@ -51,12 +53,17 @@ public DropDatabaseAnalyzer(QueryState queryState) throws SemanticException {
+      throw new SemanticException(ErrorMsg.CATALOG_NOT_EXISTS, catalogName);
+    }
+    String databaseName = catDbNamePair.getRight();
+    Database database = getDatabase(catDbNamePair.getLeft(), catDbNamePair.getRight(), !ifExists);
```
Fixed in https://github.com/apache/hive/pull/6088/commits/00a75f9fe1d13cc9e49a00a3cd2b598dadef1f0b

zhangbutao commented on ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java:
```diff
@@ -6463,6 +6564,18 @@ public List<Function> getFunctionsInDb(String dbName, String pattern) throws Hiv
     }
   }
 
+  public List<Function> getFunctionsInDb(String catName, String dbName, String pattern) throws HiveException {
```
Fixed in https://github.com/apache/hive/pull/6088/commits/cca5bdcfad23fa3653c3c51129c48dc21cce459d

zhangbutao commented on ql/src/test/queries/clientpositive/catalog.q:
```diff
@@ -45,3 +45,6 @@ SHOW CATALOGS LIKE 'test__';
 DESC CATALOG EXTENDED test_cat;
+
+-- DROP catalog at the end
+DROP CATALOG  test_cat;
```
Fixed in https://github.com/apache/hive/pull/6088/commits/00a75f9fe1d13cc9e49a00a3cd2b598dadef1f0b

zhangbutao commented:
@dengzhhu653 Thanks for the suggesstion!

> It seems to me this is a significant/big change, and the user should follow a new way of catalog.db.table to request the table as the Trino/Presto does today.

That doesn't seem quite right. My goal is that even if users don't use this feature, they can still use the classic Hive approach, such as `select * from table` or `select * from db.table`. Users don't need to worry about catalogs because their current session has a default catalog. Of course, users can also specify a catalog name as a prefix to query a table, for example, `select * from hive.db.tbl`. 

Additionally, if users want to perform federated queries across multiple data sources in Hive, similar to Trino, they would need to create multiple catalogs and execute similar queries. This federated query capability is a new feature and does not pose compatibility issues for users.
```
SELECT 
    u.user_name,
    a.action_time
FROM mysql.test.users u
JOIN hive.default.user_actions a 
    ON u.user_id = a.user_id;
```


> I think we can have a more simple idea, let's push the catalog awareness down into the Metastore client. For example,
> SET CATALOG testcat; -> change to a metaconf setting: set metaconf:metastore.catalog.default = testcat, default hive, this configuration will be propagated to HMS, every HMS API request will be appended with the testcat, both client and server can handle it.

Although this allows switching catalogs, the current client can only access the databases and tables of one catalog at a time.



> For cross-catalog queries, we can also do it in a similar way, we can introduce a new catalog awareness lawyer between session client and the thrift client, as SessionHiveMetaStoreClient -> CatalogAwarenessMetastoreClient -> ThriftHiveMetaStoreClient. The CatalogAwarenessMetastoreClient can acknowledge of the catalog through configuration or properties stored in HMS([HIVE-27186](https://issues.apache.org/jira/browse/HIVE-27186)). Something as hive.matastore.catalog.mapping.pattern=iceberg_catalog:ice_db1*,hive_db1,hive_db2.ice_tab*;hive:default,acid_db*, the CatalogAwarenessMetastoreClient even can point to a third-party catalog partner.

This method allows accessing databases and tables across multiple catalogs, but it is quite inelegant. Moreover, as you mentioned, it cannot support having the same database names across different catalogs. This solution is somewhat similar to https://github.com/ExpediaGroup/waggle-dance. Additionally, it requires users to constantly modify parameters to add or remove database names in `hive.matastore.catalog.mapping.pattern`, which seems very cumbersome...


Finally, I want to reference the multi/federated catalog capabilities of Trino and StarRocks/Doris. I believe their design approaches are quite reasonable. I'm considering implementing this capability with the principle of minimizing code modifications. Considerable preliminary work has already been done on the HMS side with HIVE-18685. The subsequent work mainly involves connecting all catalog capabilities from the HS2 side to the HMS side. I've been thinking about writing down the related design considerations, but haven't had enough time yet... I will provide more thoughts on this later.

zhangbutao commented:
@deniskuzZ Thank you very much for your review. I've been quite busy with various matters recently, which has caused some delay in addressing these comments. My apologies.

deniskuzZ commented:
> @deniskuzZ Thank you very much for your review. I've been quite busy with various matters recently, which has caused some delay in addressing these comments. My apologies.

nw at all. thank you @zhangbutao for driving this feature. I think it would be a game-changer

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/lockmgr/HiveTxnManagerImpl.java:
```diff
@@ -38,8 +39,10 @@
+import org.apache.hadoop.hive.ql.metadata.HiveUtils;
 import org.apache.hadoop.hive.ql.metadata.Partition;
 import org.apache.hadoop.hive.ql.metadata.Table;
+import org.stringtemplate.v4.ST;
```
redundant import

deniskuzZ commented on parser/src/java/org/apache/hadoop/hive/ql/parse/HiveParser.g:
```diff
@@ -379,10 +379,13 @@ TOK_CATALOGLOCATION;
 TOK_DATABASELOCATION;
 TOK_DATABASE_MANAGEDLOCATION;
 TOK_PROPLIST;
+TOK_DBPROPLIST;
```
where is that used?

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/parse/repl/load/message/CreateDatabaseHandler.java:
```diff
@@ -55,22 +55,23 @@ public List<Task<?>> handle(Context context)
     Database db = metaData.getDatabase();
     String destinationDBName =
         context.dbName == null ? db.getName() : context.dbName;
+    String destinationCatalogName = db.getCatalogName(); // TODO catalog. Need to double check the catalog here. Depend on HIVE-29278
```
could we shorten `destinationCatalogName` -> `destCatalogName`. same for destDBName

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/parse/BaseSemanticAnalyzer.java:
```diff
@@ -421,6 +421,33 @@ public static String charSetString(String charSetName, String charSetString)
+   * @return Return a Pair object which includes catalogName and dbName
+   * @throws SemanticException
+   */
+  public static Pair<String, String> getCatDbNamePair(ASTNode dbNameNode) throws SemanticException {
```
it doesn't belong in SA, maybe move to util class
